system/lib/libc/crt1.c

```diff
@@ -19,54 +19,18 @@ extern void __wasm_call_ctors(void) __attribute__((weak));
 
 // TODO(sbc): We shouldn't even link this file if there is no main:
 // https://github.com/emscripten-core/emscripten/issues/9640
-extern int main(int argc, char** argv) __attribute__((weak));
-
-// Avoid using stack allocation for argument-passed values in _start(), as the
-// stack allocation for them can't be eliminated by Binaryen later, so if we
-// have no main we end up with a stack push and pop for no reason. Also, the
-// stack allocation here would last for the entire program anyhow, so it's
-// effectively static.
-static size_t argc;
-static size_t argv_buf_size;
+extern int __original_main(void) __attribute__((weak));
 
 void _start(void) {
-  if (!main) {
-    if (__wasm_call_ctors) {
-      __wasm_call_ctors();
-    }
-    return;
-  }
-
-  /* Fill in the arguments from WASI syscalls. */
-  char **argv;
-  __wasi_errno_t err;
-
-  /* Get the sizes of the arrays we'll have to create to copy in the args. */
-  err = __wasi_args_sizes_get(&argc, &argv_buf_size);
-  if (err != __WASI_ESUCCESS) {
-    __wasi_proc_exit(EX_OSERR);
-  }
-
-  if (argc) {
-    /* Allocate memory for the array of pointers, adding null terminator. */
-    argv = alloca(sizeof(char *) * (argc + 1));
-    /* Allocate memory for storing the argument chars. */
-    char *argv_buf = alloca(sizeof(char) * argv_buf_size);
-    /* Make sure the last pointer in the array is NULL. */
-    argv[argc] = NULL;
-    /* Fill the argument chars, and the argv array with pointers into those chars. */
-    err = __wasi_args_get(argv, argv_buf);
-    if (err != __WASI_ESUCCESS) {
-      __wasi_proc_exit(EX_OSERR);
-    }
-  }
-
-  /* The linker synthesizes this to call constructors. */
   if (__wasm_call_ctors) {
     __wasm_call_ctors();
   }
 
-  int r = main(argc, argv);
+  if (!__original_main) {
```

Comment:
Does this double level of weak symbols work?  Is it needed?  Its making my head hurt a little trying to reason about it.

Reply:
I see on CI that this does seem to fail on the case where there is no main at all. Is that what you meant?

I don't know much about weak symbols :) Is there a better way to do all this?

Reply:
I pushed an update that handles the case without main. We need to check if main exists in both `_start` and `__original_main`, basically (by doing so in `__original_main` we can let the optimizer remove the wasi args code).